File: Current/Product/Production/Common/Common.Tasks/XmlQuery.vb
Option Explicit On
Option Strict On

Imports System.IO
Imports System.Xml
Imports NAnt.Core
Imports NAnt.Core.Types
Imports NAnt.Core.Attributes

<ElementName("xmlquery")> _
Public Class XmlQuery
    Inherits LoopItems

    Private _Query As String
    Private _File As String
    Private _Namespaces As XmlNamespaceCollection


    <BuildElementCollection("namespaces", "namespace")> _
    Public Property Namespaces() As XmlNamespaceCollection
        Get
            Return Me._Namespaces
        End Get
        Set(ByVal value As XmlNamespaceCollection)
            Me._Namespaces = value
        End Set
    End Property

    <TaskAttribute("query", Required:=True)> _
    Public Property Query() As String
        Get
            Return _Query
        End Get
        Set(ByVal value As String)
            _Query = value
        End Set
    End Property

    <TaskAttribute("file", Required:=True)> _
    Public Property File() As String
        Get
            Return _File
        End Get
        Set(ByVal value As String)
            _File = value
        End Set
    End Property

    Protected Overrides Function GetStrings() As System.Collections.IEnumerator
        Dim Strings As New ArrayList
        Dim XmlDoc As New XmlDocument
        XmlDoc.Load(Me.File)

        Dim Manager As New XmlNamespaceManager(XmlDoc.NameTable)
        Dim [NameSpace] As XmlNamespace
        For Each [NameSpace] In Me.Namespaces
            If ([NameSpace].IfDefined AndAlso Not [NameSpace].UnlessDefined) Then
                Manager.AddNamespace([NameSpace].Prefix, [NameSpace].Uri)
            End If
        Next

        For Each Node As XmlNode In XmlDoc.SelectNodes(Me.Query, Manager)
            Strings.Add(Node.InnerXml)
        Next

        Return DirectCast(Strings.ToArray(GetType(String)), String()).GetEnumerator
    End Function

End Class

File: tags/1.0.0.50/Product/Production/Nant/CIFactory.NAnt.Tasks/Types/XmlQuery.cs
using System;
using System.Xml;
using System.Collections;
using NAnt.Core.Types;
using NAnt.Core.Attributes;

namespace CIFactory.NAnt.Types
{
    [ElementName("xmlquery")]
    public class XmlQuery : LoopItems
    {
        #region Fields

        private string _File;

        private XmlNamespaceCollection _Namespaces;

        private string _Query;

        #endregion

        #region Properties

        [TaskAttribute("file", Required = true)]
        public string File
        {
            get { return _File; }
            set { _File = value; }
        }

        [BuildElementCollection("namespaces", "namespace")]
        public XmlNamespaceCollection Namespaces
        {
            get { return this._Namespaces; }
            set { this._Namespaces = value; }
        }

        [TaskAttribute("query", Required = true)]
        public string Query
        {
            get { return _Query; }
            set { _Query = value; }
        }

        #endregion

        #region Protected Methods

        protected override System.Collections.IEnumerator GetStrings()
        {
            ArrayList Strings = new ArrayList();
            XmlDocument XmlDoc = new XmlDocument();
            XmlDoc.Load(this.File);
            XmlNamespaceManager Manager = new XmlNamespaceManager(XmlDoc.NameTable);

            foreach (XmlNamespace NameSpace in this.Namespaces)
            {
                if ((NameSpace.IfDefined && !NameSpace.UnlessDefined))
                {
                    Manager.AddNamespace(NameSpace.Prefix, NameSpace.Uri);
                }
            }
            foreach (XmlNode Node in XmlDoc.SelectNodes(this.Query, Manager))
            {
                if (Node.Value != null)
                    Strings.Add(Node.Value);
                else
                    Strings.Add(Node.InnerXml);
            }
            return ((string[])Strings.ToArray(typeof(string))).GetEnumerator();
        }

        #endregion

    }
}